PR: https://github.com/odoo/odoo/pull/

From: 61eef73b52c078daaae37f8d57e93830a2e53410
From: Christophe Matthieu
Date: 2018-08-13 15:09:28

Structural Changes: 4
Total Changes: 70

[IMP] base, *: use a many2one for the company report layout

Before this rev. one could set the report layout on the company using a
hardcoded list (background, clean, standard, etc.). Other modules (typically
accounting modules) could add options in this list. The used template was built
on the layout key.

Now, the layout is a many2one field to a newly created model `report.layout`,
which is linked to a view with the layout architecture.

This gives more control to customize reports and create a new layout (without
creating a python module that extend the layout selection).

================================= pseudo patch: =================================

--- a/odoo/addons/base/data/res_company_demo.xml
+++ b/odoo/addons/base/data/res_company_demo.xml
@@ -3,7 +3,6 @@
 <data noupdate="1">
     <record id="main_company" model="res.company">
         <field name="name">ProBike Inc</field>
-        <field name="external_report_layout">standard</field>
     </record>
 </data>
 </odoo>

--- a/odoo/addons/base/models/ir_actions_report.py
+++ b/odoo/addons/base/models/ir_actions_report.py
@@ -710,7 +710,7 @@ class IrActionsReport(models.Model):
         :param report_name: Name of the template to generate an action for
         """
         discard_logo_check = self.env.context.get('discard_logo_check')
-        if (self.env.uid == SUPERUSER_ID) and ((not self.env.user.company_id.external_report_layout) or (not discard_logo_check and not self.env.user.company_id.logo)) and config:
+        if (self.env.uid == SUPERUSER_ID) and ((not self.env.user.company_id.external_report_layout_id) or (not discard_logo_check and not self.env.user.company_id.logo)) and config:
             template = self.env.ref('base.view_company_report_form_with_print') if self.env.context.get('from_transient_model', False) else self.env.ref('base.view_company_report_form')
             return {
                 'name': _('Choose Your Document Layout'),

--- a/odoo/addons/base/models/ir_qweb_fields.py
+++ b/odoo/addons/base/models/ir_qweb_fields.py
@@ -655,6 +655,4 @@ class QwebView(models.AbstractModel):
             _logger.warning("%s.%s must be a 'ir.ui.view' model." % (record, field_name))
             return None
 
-        view = view.with_context(object=record)
-
-        return pycompat.to_text(view.render(view._context, engine='ir.qweb'))
+        return pycompat.to_text(view.render(options.get('values', {}), engine='ir.qweb'))

--- a/odoo/addons/base/models/res_company.py
+++ b/odoo/addons/base/models/res_company.py
@@ -55,14 +55,8 @@ class Company(models.Model):
     website = fields.Char(related='partner_id.website')
     vat = fields.Char(related='partner_id.vat', string="Tax ID")
     company_registry = fields.Char()
-    paperformat_id = fields.Many2one('report.paperformat', 'Paper Format', default=lambda self: self.env.ref('base.paperformat_euro', raise_if_not_found=False))
-    external_report_layout = fields.Selection([
-        ('background', 'Background'),
-        ('boxed', 'Boxed'),
-        ('clean', 'Clean'),
-        ('standard', 'Standard'),
-    ], string='Document Template')
-
+    paperformat_id = fields.Many2one('report.paperformat', 'Paper format', default=lambda self: self.env.ref('base.paperformat_euro', raise_if_not_found=False))
+    external_report_layout_id = fields.Many2one('ir.ui.view', 'Document Template')
     _sql_constraints = [
         ('name_uniq', 'unique (name)', 'The company name must be unique !')
     ]

--- a/odoo/addons/base/static/img/preview_background.png
+++ b/odoo/addons/base/static/img/preview_background.png
Binary files a/odoo/addons/base/static/img/preview_background.png and /dev/null differ

--- a/odoo/addons/base/static/img/preview_boxed.png
+++ b/odoo/addons/base/static/img/preview_boxed.png
Binary files a/odoo/addons/base/static/img/preview_boxed.png and /dev/null differ

--- a/odoo/addons/base/static/img/preview_clean.png
+++ b/odoo/addons/base/static/img/preview_clean.png
Binary files a/odoo/addons/base/static/img/preview_clean.png and /dev/null differ

--- a/odoo/addons/base/static/img/preview_standard.png
+++ b/odoo/addons/base/static/img/preview_standard.png
Binary files a/odoo/addons/base/static/img/preview_standard.png and /dev/null differ

--- a/odoo/addons/base/static/pdf/preview_background.pdf
+++ b/odoo/addons/base/static/pdf/preview_background.pdf
Binary files a/odoo/addons/base/static/pdf/preview_background.pdf and /dev/null differ

--- a/odoo/addons/base/static/pdf/preview_boxed.pdf
+++ b/odoo/addons/base/static/pdf/preview_boxed.pdf
Binary files a/odoo/addons/base/static/pdf/preview_boxed.pdf and /dev/null differ

--- a/odoo/addons/base/static/pdf/preview_clean.pdf
+++ b/odoo/addons/base/static/pdf/preview_clean.pdf
Binary files a/odoo/addons/base/static/pdf/preview_clean.pdf and /dev/null differ

--- a/odoo/addons/base/static/pdf/preview_standard.pdf
+++ b/odoo/addons/base/static/pdf/preview_standard.pdf
Binary files a/odoo/addons/base/static/pdf/preview_standard.pdf and /dev/null differ

--- a/odoo/addons/base/views/res_company_views.xml
+++ b/odoo/addons/base/views/res_company_views.xml
@@ -141,23 +141,9 @@
               <xpath expr="//sheet" position="inside">
                     <hr />
                     <div class="row mt16">
-                        <field style="margin-right: 40px;" name="external_report_layout" widget="image_selection" options="{
-                            'background': {
-                                'image_link': '/base/static/img/preview_background.png',
-                                'preview_link': '/base/static/pdf/preview_background.pdf'
-                            },
-                            'boxed': {
-                                'image_link': '/base/static/img/preview_boxed.png',
-                                'preview_link': '/base/static/pdf/preview_boxed.pdf'
-                            },
-                            'clean': {
-                                'image_link': '/base/static/img/preview_clean.png',
-                                'preview_link': '/base/static/pdf/preview_clean.pdf'
-                            },
-                            'standard': {
-                                'image_link': '/base/static/img/preview_standard.png',
-                                'preview_link': '/base/static/pdf/preview_standard.pdf'
-                            }
+                        <field style="margin-right: 40px;" name="external_report_layout_id" widget="report_layout" options="{
+                            'field_image': 'preview_image',
+                            'field_binary': 'preview_pdf'
                         }"/>
                     </div>
                     <footer>
@@ -187,32 +173,21 @@
             <field name="priority">2048</field>
             <field name="arch" type="xml">
                 <form class="o_company_document_layout">
+                    <group>
+                        <label for="external_report_layout_id" string="Template" colspan="2" />
+                        <field name="external_report_layout_id" colspan="2" nolabel="1"
+                                class="report_layout_container"
+                                widget="report_layout" options="{
+                                    'field_image': 'preview_image',
+                                    'field_binary': 'preview_pdf'
+                                }"/>
+                    </group>
                     <group>
                         <field name="logo" widget="image" />
-                        <field name="external_report_layout" string="Layout"
-                               class="image_selection_container"
-                               widget="image_selection" options="{
-                                    'background': {
-                                        'image_link': '/base/static/img/preview_background.png',
-                                        'preview_link': '/base/static/pdf/preview_background.pdf'
-                                    },
-                                    'boxed': {
-                                        'image_link': '/base/static/img/preview_boxed.png',
-                                        'preview_link': '/base/static/pdf/preview_boxed.pdf'
-                                    },
-                                    'clean': {
-                                        'image_link': '/base/static/img/preview_clean.png',
-                                        'preview_link': '/base/static/pdf/preview_clean.pdf'
-                                    },
-                                    'standard': {
-                                        'image_link': '/base/static/img/preview_standard.png',
-                                        'preview_link': '/base/static/pdf/preview_standard.pdf'
-                                    }
-                                }" />
                         <field name="report_header"  widget="char"
-                               placeholder="e.g. Global Business Solutions" />
+                            placeholder="e.g. Global Business Solutions" />
                         <field name="report_footer" string="Footer"
-                               placeholder="e.g. Opening hours, bank accounts (one per line)" />
+                            placeholder="e.g. Opening hours, bank accounts (one per line)" />
                         <field name="paperformat_id" widget="selection" />
                     </group>
                     <footer>
